File: controllers/chatController.js
import Groq from "groq-sdk";
import Chat from "../models/Chat.js";

export const sendMessage = async (req, res) => {
  try {
    const groq = new Groq({
      apiKey: process.env.GROQ_API_KEY
    });

    const { message } = req.body;
    const userId = req.userId;

    if (!message || message.trim() === "") {
      return res.status(400).json({
        success: false,
        message: "Message cannot be empty"
      });
    }

    // Get previous chat history (last 10 messages for context)
    const previousChats = await Chat.find({ userId })
      .sort({ timestamp: -1 })
      .limit(10)
      .lean();

    // Reverse to get chronological order
    previousChats.reverse();

    // Build conversation history for AI
    const conversationHistory = [
      {
        role: "system",
        content: `You are an expert Indian fitness and nutrition coach named "SwasthAI Coach". 
        
Your expertise includes:
- Indian diet and nutrition
- Fitness and exercise guidance
- Health management for Indian lifestyle
- Ayurvedic wellness tips

Guidelines:
- Give practical, actionable advice
- Use Indian food examples (roti, dal, sabzi, etc.)
- Keep responses concise (2-3 sentences for simple queries)
- Be encouraging and motivating
- Suggest Indian home remedies when appropriate
- Always prioritize safety - suggest doctor consultation for serious issues
- Remember previous conversation context and refer to it when relevant`
      }
    ];

    // Add previous chat history
    previousChats.forEach(chat => {
      conversationHistory.push({
        role: chat.role === "user" ? "user" : "assistant",
        content: chat.message
      });
    });

    // Add current user message
    conversationHistory.push({
      role: "user",
      content: message
    });

    const chatCompletion = await groq.chat.completions.create({
      messages: conversationHistory,
      model: "llama-3.3-70b-versatile",
      temperature: 0.8,
      max_tokens: 500,
    });

    const reply = chatCompletion.choices[0]?.message?.content || "Sorry, I couldn't generate a response. Please try again.";

    // Save user message to database
    await Chat.create({
      userId,
      role: "user",
      message: message.trim()
    });

    // Save AI response to database
    await Chat.create({
      userId,
      role: "assistant",
      message: reply.trim()
    });

    res.status(200).json({
      success: true,
      reply: reply.trim()
    });

  } catch (error) {
    console.error("Chat Error:", error.message);
    
    // Handle specific Groq errors
    if (error.message.includes("API key")) {
      return res.status(401).json({
        success: false,
        message: "API configuration error. Please contact support.",
        error: "Invalid API credentials"
      });
    }

    res.status(500).json({ 
      success: false,
      message: "Chat service temporarily unavailable",
      error: error.message 
    });
  }
};

// Get chat history
export const getChatHistory = async (req, res) => {
  try {
    const userId = req.userId;
    
    const chats = await Chat.find({ userId })
      .sort({ timestamp: 1 })
      .limit(50)
      .lean();

    res.status(200).json({
      success: true,
      chats
    });

  } catch (error) {
    console.error("Get Chat History Error:", error.message);
    res.status(500).json({ 
      success: false,
      message: "Failed to fetch chat history",
      error: error.message 
    });
  }
};

// Clear chat history
export const clearChatHistory = async (req, res) => {
  try {
    const userId = req.userId;
    
    await Chat.deleteMany({ userId });

    res.status(200).json({
      success: true,
      message: "Chat history cleared successfully"
    });

  } catch (error) {
    console.error("Clear Chat History Error:", error.message);
    res.status(500).json({ 
      success: false,
      message: "Failed to clear chat history",
      error: error.message 
    });
  }
};

File: controllers/chat_controller.py
import os
from fastapi import HTTPException
from pydantic import BaseModel
from groq import Groq
from models.chat import Chat
from dotenv import load_dotenv

load_dotenv()


# ================= REQUEST SCHEMAS =================

class SendMessageRequest(BaseModel):
    message: str


# ================= SEND MESSAGE =================

async def send_message(data: SendMessageRequest, user_id: str):
    try:
        if not data.message or not data.message.strip():
            raise HTTPException(status_code=400, detail="Message cannot be empty")

        groq_client = Groq(api_key=os.getenv("GROQ_API_KEY"))

        # Last 10 messages fetch karo context ke liye
        previous_chats = await Chat.find(
            Chat.userId == user_id
        ).sort(-Chat.timestamp).limit(10).to_list()

        # Chronological order mein reverse karo
        previous_chats.reverse()

        # Conversation history build karo
        conversation_history = [
            {
                "role": "system",
                "content": """You are an expert Indian fitness and nutrition coach named "SwasthAI Coach". 
        
Your expertise includes:
- Indian diet and nutrition
- Fitness and exercise guidance
- Health management for Indian lifestyle
- Ayurvedic wellness tips

Guidelines:
- Give practical, actionable advice
- Use Indian food examples (roti, dal, sabzi, etc.)
- Keep responses concise (2-3 sentences for simple queries)
- Be encouraging and motivating
- Suggest Indian home remedies when appropriate
- Always prioritize safety - suggest doctor consultation for serious issues
- Remember previous conversation context and refer to it when relevant"""
            }
        ]

        for chat in previous_chats:
            conversation_history.append({
                "role": "user" if chat.role == "user" else "assistant",
                "content": chat.message
            })

        conversation_history.append({
            "role": "user",
            "content": data.message
        })

        # Groq API call
        chat_completion = groq_client.chat.completions.create(
            messages=conversation_history,
            model="llama-3.3-70b-versatile",
            temperature=0.8,
            max_tokens=500,
        )

        reply = chat_completion.choices[0].message.content or "Sorry, I couldn't generate a response. Please try again."

        # User message save karo
        await Chat(
            userId=user_id,
            role="user",
            message=data.message.strip()
        ).insert()

        # AI response save karo
        await Chat(
            userId=user_id,
            role="assistant",
            message=reply.strip()
        ).insert()

        return {"success": True, "reply": reply.strip()}

    except HTTPException:
        raise
    except Exception as e:
        error_msg = str(e)
        if "API key" in error_msg:
            raise HTTPException(status_code=401, detail="API configuration error. Please contact support.")
        raise HTTPException(status_code=500, detail=f"Chat service temporarily unavailable: {error_msg}")


# ================= GET CHAT HISTORY =================

async def get_chat_history(user_id: str):
    try:
        chats = await Chat.find(
            Chat.userId == user_id
        ).sort(+Chat.timestamp).limit(50).to_list()

        chats_data = []
        for chat in chats:
            c = chat.dict()
            c["_id"] = str(chat.id)
            chats_data.append(c)

        return {"success": True, "chats": chats_data}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch chat history: {str(e)}")


# ================= CLEAR CHAT HISTORY =================

async def clear_chat_history(user_id: str):
    try:
        await Chat.find(Chat.userId == user_id).delete()
        return {"success": True, "message": "Chat history cleared successfully"}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to clear chat history: {str(e)}")
